### DISCUSSION TITLE
EnjoyApp无法添加时长超过1小时的本地音频资源

### DISCUSSION THREAD

**Discussion by eatgrass** [ZuodaoTech/everyone-can-use-english#720] [Q&A] [answered]
![image](https://github.com/xiaolai/everyone-can-use-english/assets/2351076/5a9d739f-8483-47dd-aaea-e10e5f226590)

如题。

**Comment by an-lee** [comment]
音视频太大的时候，在生成声波图形时很容易卡死。目前对单条音频的大小做了限制，超过 50M 的会导入失败。
如果是音频书，建议先按章节做切割再使用。

**Reply by eatgrass** [reply]
并非有声书，而是像一些播客节目，一期就是会很长。

比如
https://podcasts.apple.com/us/podcast/sean-carrolls-mindscape-science-society-philosophy/id1406534739
https://podcasts.apple.com/us/podcast/lex-fridman-podcast/id1434243584

另外非常感谢你们开发这个App, 真的非常好

**Comment by czyrichard** [comment]
用Audacity打开音频，导出音频为mp3格式，比特率模式选“可变”，质量选“65-105kbps”，再勾选“强制输出到单声道”，这样20分钟的音频可以压缩到6MB左右，这样哪怕是两个小时的音频应该也能导入了。

**Reply by eatgrass** [reply]
我使用ffmpeg压缩后，确实能导进去了，但是如 @an-lee 所说，会卡死
![image](https://github.com/ZuodaoTech/everyone-can-use-english/assets/2351076/9676b69d-db5e-4a8c-88a4-5485765262b5)

**Reply by an-lee** [reply, marked as answer]
实际不是文件大小的问题，是音频时长的问题，目前只是临时用文件大小做了限制。

后续优化有两个思路：

1. 自动分割文件，分段读取
2. waveform 分段渲染

目前建议用 ffmpeg 切割之后，再导入吧。

**Reply by jemerci** [reply]
我也卡死了, 约90 分钟的音频, 10M 大小左右, 没看到这个 thread , 折腾了一个晚上. 还换了一台高配的win机器, same
win10专业版, v0.2.14 Enjoy.

**Reply by tan2meng** [reply]
> 我也卡死了, 约90 分钟的音频, 10M 大小左右, 没看到这个 thread , 折腾了一个晚上. 还换了一台高配的win机器, same win10专业版, v0.2.14 Enjoy.

用[LosslessCut](https://github.com/mifi/lossless-cut)剪非常方便，一次把几小时的podcast剪成数个小段。

**Comment by czyrichard** [comment]
我让ChatGPT给我写了两个脚本，先把这个脚本和需要处理的mp3文件放在同一目录下，然后逐一运行就可以啦：

第一个脚本（"生成该目录下所有mp3文件的静音部分的时间戳.py"）可以自动检测音频中所有的静音片段，将其记录在一个txt文件中；

第二个脚本（"根据mp3静音时间戳切割mp3为2分钟左右片段.py"）可以根据txt文件提供的时间戳计算出该在哪里对音频进行切割，将MP3文件切割成若干个两分钟左右的mp3文件，并保证切割的时候不会把一个句子从中间切开。

[切割有声书专用脚本.zip](https://github.com/user-attachments/files/16241118/default.zip)
